### PR TITLE
SMCGT-1765 - Database loss

### PR DESCRIPTION
**Pull request by justin-koevort**
- SMCGT-1765 - Fix database loss after watchdog restart (FieldServer Mk-ii)

### CODE REVIEW

**Review by JensEggers — DISMISSED** [review]

**Comment by JensEggers** [comment]
@justin-koevort you may need to talk to @southbite (Simon) to figure out dependency updates before merging

File: packages/happn-db-provider-loki/index.js [hunk 1]
```diff
@@ -720,6 +726,8 @@ module.exports = class LokiDataProvider extends commons.BaseDataProvider {
             callback(errorSyncing);
             return;
           }
+          if (fs.existsSync(this.settings.tempDataFilename))
```
[the remark this answers is not in the document]
Comment (JensEggers) [reply packages/happn-db-provider-loki/index.js:729]: sometimes half that, sometimes more

[the remark this answers is not in the document]
Comment (JensEggers) [reply packages/happn-db-provider-loki/index.js:729]: so we're fine in terms of impact on the event loop

File: packages/happn-db-provider-loki/index.js [hunk 2]
```diff
@@ -828,4 +836,16 @@ module.exports = class LokiDataProvider extends commons.BaseDataProvider {
       throw new Error('Loaded archives only support read-only operations!');
     }
   }
+
+  #getFileSize(filepath) {
+    let fileStats = null;
+    try {
+      fileStats = fs.statSync(filepath);
```
[the remark this answers is not in the document]
Comment (JensEggers) [reply packages/happn-db-provider-loki/index.js:843]: again, we're fine in terms of impact on the event loop

[the remark this answers is not in the document]
Comment (justin-koevort) [reply packages/happn-db-provider-loki/index.js:843]: Any chance unlinkSync may be an issue?

File: packages/happn-db-provider-loki/index.js [hunk 3]
```diff
@@ -607,14 +607,18 @@ module.exports = class LokiDataProvider extends commons.BaseDataProvider {
...
+    fs.rename(this.settings.tempDataFilename, this.settings.filename, callback);
   }
 
   copyMainDataToArchive(suffix, callback) {
     if (fs.existsSync(`${this.settings.filename}.${suffix}`))
-      fs.unlinkSync(`${this.settings.filename}.${suffix}`);
+      fs.unlink(`${this.settings.filename}.${suffix}`, (err) => {
+        if (err) this.logger.error(err.message);
+      });
     fs.copy(this.settings.filename, `${this.settings.filename}.${suffix}`, callback);
```
Comment (JensEggers) [inline packages/happn-db-provider-loki/index.js:622]: this looks suspect - you're not waiting for unlink to complete (if it needs to) before copying the file, so we have a race condition. Best would be to have a callback-based function that verifies that the old file does not exist (by deleting if necessary), and only starting the copy in its callback.

**Review by JensEggers — DISMISSED** [review]

**Review by riaannieuwoudt — APPROVED** [review]

File: packages/happn-db-provider-loki/index.js (hunk 3 above)
Comment (southbite) [reply packages/happn-db-provider-loki/index.js]: unlinkSync is a synchronous operation, copy will never be called during this operation.

File: packages/happn-db-provider-loki/index.js [hunk 4]
```diff
@@ -828,4 +843,16 @@ module.exports = class LokiDataProvider extends commons.BaseDataProvider {
       throw new Error('Loaded archives only support read-only operations!');
     }
   }
+
+  #getFileSize(filepath) {
+    let fileStats = null;
+    try {
+      fileStats = fs.statSync(filepath);
```
Comment (csampsonza) [inline packages/happn-db-provider-loki/index.js:850]: Why are we doing this all as Sync function?  we should rather just put locks around changing the DB while we are storing and not block other async IO.  

I think calculating size, will be very depend on the file size, calculation on a couple of Megs will take a while.

Comment (JensEggers) [reply packages/happn-db-provider-loki/index.js:850]: we took a quantitative look at this - it's a native Node.js operation relying on the cached file system, and measured it to take < 1ms regardless of file size. So the cost is negligible compared to real-life usage patterns and does not merit further optimization.

**Review by csampsonza — APPROVED** [review]

File: packages/happn-db-provider-loki/index.js [hunk 5]
```diff
@@ -69,6 +71,7 @@ module.exports = class LokiDataProvider extends commons.BaseDataProvider {
     }
 
     this.settings.snapshotRollOverThreshold = this.settings.snapshotRollOverThreshold || 1e3; // take a snapshot and compact every 1000 records
+    this.settings.fileSizeDifferenceLimit = this.settings.fileSizeDifferenceLimit || 1e7;
```
Comment (csampsonza) [inline packages/happn-db-provider-loki/index.js:74]: I think we should not do a breaking change here.  But this feature and default to the same operation as before.  i.e. off, and no checking file size on every storePlayback.  Then via configuration we can swich this on.

File: packages/happn-db-provider-loki/index.js [hunk 6]
```diff
@@ -628,9 +627,18 @@ module.exports = class LokiDataProvider extends commons.BaseDataProvider {
...
           return callback(appendFailure);
         }
-        if (this.operationCount < this.settings.snapshotRollOverThreshold) {
+        let currentSize = 0;
+        if (this.settings.fileSizeDifferenceLimit > 0) {
+          currentSize = this.#getFileSize(this.settings.filename);
+        }
+        if (
+          this.operationCount < this.settings.snapshotRollOverThreshold &&
+          (this.settings.fileSizeDifferenceLimit === 0 ||
```
Comment (csampsonza) [inline packages/happn-db-provider-loki/index.js:636]: Rather use

if (  currentSize === 0 || currentSize - this.baselineFileSize < this.settings.fileSizeDifferenceLimit )

File: packages/happn-db-provider-loki/index.js (hunk 5 above)
Comment (csampsonza) [reply packages/happn-db-provider-loki/index.js]: Fixed, default to not checking for this.

**Review by csampsonza — APPROVED** [review]

**Review by southbite — APPROVED** [review]

**Pull request merged**